scripts/dts: globals: parse gpio-map nodes

erwango opened the pull request:
gpio-map is defined in dts v0.3 as "nexus node".
It allows to describe a pin connector so it can be referenced
through phandles and hence used in expansion device nodes
(typically implemented through overlays).

This change implements gpio controller resolution through these maps.
Few assumptions were taken in order to simplify the implementation.
These assumptions enforce some limitations to the use of gpio-map
but my understanding is that this should still allow to cover most
use cases.

Assumptions:
-gpio-size is the same for all gpio-controllers referenced in a map
-optional properties gpio-map-mask and gpio-map-pass-thru are
supposed to be omitted
The understanding of this last assumption is that flags provided in
the expansion device node will overwrite the connector flags

Fixes #15637

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@mike-scott, I had a try on a (former) modified version of your PR and it worked. Can you have a try when you have some time?

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +332,30 @@ def build_cell_array(prop_array):
+    index = 0
+    map_array = []
+
+    raw_map = reduced[cell_parent]['props'].get('gpio-map')
```
Using `.get('gpio-map')` instead of `['gpio-map']` here isn't safer, because if `raw_map` ever ends up as `None`, then `len(raw_map)` below will crash.

erwango replied:
Ok, so using ['gpio-map'], then.
This should not end up as `None` since function is called only if `gpio-map` in `reduced...`.

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +332,30 @@ def build_cell_array(prop_array):
+    # child specifier field lenght is connector #gpio-cells
+    # parent specifier field lenght is parent #gpio-cells
+
+    child_specifier_size = reduced[cell_parent]['props'].get('#gpio-cells')
```
Ditto here, re. `.get()` vs. `[]`.

erwango replied:
ack

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +332,30 @@ def build_cell_array(prop_array):
+    # TODO: for now we assume parent #gpio-cells is constant across the map
+    # so we take the first one and apply in the whole map
+    parent = phandles[child_specifier_size+1]
+    parent_specifier_size = reduced[parent]['props'].get('#gpio-cells')
```
Ditto here, re. `.get()` vs. `[]`.

erwango replied:
ack

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -350,6 +375,15 @@ def extract_controller(node_path, prop, prop_values, index,
+            try:
+                # controller is to be found in the map, using elem[1] as index
+                cell_parent = gpio_map_lookup(cell_parent, elem[1])
+            except:
```
A bare `try: ... except:` is usually a bad idea, because it will hide stuff like references to undefined Python variables as well (those throw exceptions).

Catch `Exception` instead, or (better) do a more explicit check for whatever this is checking, without exceptions.

erwango replied:
Yeah, so actually this is the result of a late refactoring and this not chacking anything good anymore. Let me rework that.

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -408,6 +442,14 @@ def extract_cells(node_path, prop, prop_values, names, index,
+                # controller is to be found in the map, using elem[1] as index
+                cell_parent = gpio_map_lookup(cell_parent, elem[1])
+            except:
+                raise Exception("No match in gpio-map")
```
Ditto here.

erwango commented on scripts/dts/extract/globals.py:
```diff
@@ -8,6 +8,7 @@
 import sys
 
 from collections import defaultdict
+from pprint import pprint
```
Oversight, txs!

erwango commented:
@ulfalizer, thanks for the review. Updated.

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,44 @@ def build_cell_array(prop_array):
 
     return ret_array
 
+def gpio_map_lookup(cell_parent, gpio_index):
```
It looks like the only thing this function uses `cell_parent` for is to look up `reduced[cell_parent]['props']`. Do you think it could just take the properties directly as an argument instead (something like `cell_parent_props` maybe)?

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,44 @@ def build_cell_array(prop_array):
+    map_array = []
+
+    # Aim of this function is to return a gpio-controller phandle
+    # from the object provided (cell_parent: gpio-map) thanks to the
```
Does the `:` mean that `cell_parent` is a `gpio-map`? Might be clearer to write it out in that case ("cell_parent, which is a gpio-map").

"From the object provided" is a bit confusing since the function takes two arguments.

erwango replied:
Reworked

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,44 @@ def build_cell_array(prop_array):
 
+def gpio_map_lookup(cell_parent, gpio_index):
+    index = 0
+    map_array = []
```
These two are only used at the end of the function. Might be clearer to move them there, right before where they're used.

Then the documentation for the function ends up at the top too.

erwango replied:
Done

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -350,6 +388,15 @@ def extract_controller(node_path, prop, prop_values, index,
+                # controller is to be found in the map, using elem[1] as index
+                cell_parent = gpio_map_lookup(cell_parent, elem[1])
+            except IndexError:
+                raise Exception("GPIO index not available in gpio-map")
```
Maybe `gpio_map_lookup()` could return `None` or something instead if the controller isn't found. Reading this quickly, you wonder if the `IndexError` is there to guard the `elem[1]` lookup or the like.

Exceptions for control flow often gets confusing.

erwango replied:
So I reworked a bit, but I'm not finding that much better in terms of exception, but seems more coherent in the dt object type handling. Let me know.

ulfalizer reviewed (changes requested):
Do you think the function could have some `x_to_y()`-like name btw, so you can tell from the name what it maps to what? From the current name, you can only tell that it's doing some kind of mapping somehow related to GPIO.

mike-scott commented:
> @mike-scott, I had a try on a (former) modified version of your PR and it worked. Can you have a try when you have some time?
> 
> EDIT: Here is a branch making it working on top of #14057. Remaining detail is the change of a _0 to _1. Otherwise it, at least, compiles. I haven't been able to test though.

Absolutely.  Thank you for working on this.  I'll respond back once I kick the tires a bit and look through this.

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,51 @@ def build_cell_array(prop_array):
+def nexus_to_controller(cell_parent, gpio_index):
+    # Aim of this function is to return a gpio-controller
+    # from the cell_parent provided which has been identified as a 'nexus node'
+    # (ie has a gpio-map property).
```
Accidental period in the middle of the sentence?

mike-scott commented:
@erwango Actually, I don't see the branch based on the shield changes?  Did I miss a link?

erwango commented:
@mike-scott, you didn't miss anything, I omited the paste operation. Here it is: https://github.com/erwango/zephyr/tree/test_14057_gpio_map

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,51 @@ def build_cell_array(prop_array):
+    try:
+        # controller phandle is to be found in the map, using gpio_index
+        controller_phandle = map_array[gpio_index][child_specifier_size]
+        return phandles[controller_phandle]
```
Can't immediately tell which of the `[gpio_index]`, `[child_specifier_size]`, and `[controller_phandle]` lookups are meant to be able to generate the `IndexError` under "normal" operation of the program here.

Better to do something like this, and avoid `try`/`except` completely:

```python
if gpio_index not in map_array:
    return None

...
```

That makes the intention clear, and means that errors in the code won't be eaten by the `except IndexError:`.

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -350,6 +395,16 @@ def extract_controller(node_path, prop, prop_values, index,
+            # parent is a gpio 'nexus node' (ie has gpio-map)
+            # controller is to be found in the map, using elem[1] as index
+            gpio_controller = nexus_to_controller(cell_parent, elem[1])
+            if gpio_controller is not 'None':
```
Should be `is not None`, not `is not 'None'`. Just `if gpio_controller:` should work though.

Something like this:

```python
if not gpio_controller:
    raise Exception("GPIO controller not found")
cell_parent = gpio_controller
```

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -408,6 +463,15 @@ def extract_cells(node_path, prop, prop_values, names, index,
+            # parent is a gpio 'nexus node' (ie has gpio-map)
+            # controller is to be found in the map, using elem[1] as index
+            gpio_controller = nexus_to_controller(cell_parent, elem[1])
+            if gpio_controller is not 'None':
```
Ditto here.

mike-scott commented:
@erwango I tested this tonight and noticed that the values for the PINs was the index of the gpio-map used (example is for Sara R4 modem -- see header pins 12 and 11):
```
		mdm-reset-gpios = <&arduino_header 12 0>;
		mdm-power-gpios = <&arduino_header 11 0>;
```

Example:
```
#define DT_NXP_KINETIS_UART_4006D000_UBLOX_SARA_R4_4006D000_SARA_R4_MDM_POWER_GPIOS_CONTROLLER		"GPIO_0"
#define DT_NXP_KINETIS_UART_4006D000_UBLOX_SARA_R4_4006D000_SARA_R4_MDM_POWER_GPIOS_FLAGS		0
#define DT_NXP_KINETIS_UART_4006D000_UBLOX_SARA_R4_4006D000_SARA_R4_MDM_POWER_GPIOS_PIN			11
#define DT_NXP_KINETIS_UART_4006D000_UBLOX_SARA_R4_4006D000_SARA_R4_MDM_RESET_GPIOS_CONTROLLER		"GPIO_2"
#define DT_NXP_KINETIS_UART_4006D000_UBLOX_SARA_R4_4006D000_SARA_R4_MDM_RESET_GPIOS_FLAGS		0
#define DT_NXP_KINETIS_UART_4006D000_UBLOX_SARA_R4_4006D000_SARA_R4_MDM_RESET_GPIOS_PIN			12
```

To fix this, I used this hacky code which updates the prop_values based on the content of the gpio-map:
```
diff --git a/scripts/dts/extract/globals.py b/scripts/dts/extract/globals.py
index 8f1097a803..a835e2fa89 100644
--- a/scripts/dts/extract/globals.py
+++ b/scripts/dts/extract/globals.py
@@ -331,7 +331,7 @@ def build_cell_array(prop_array):
 
     return ret_array
 
-def gpio_map_lookup(cell_parent, gpio_index):
+def gpio_map_lookup(cell_parent, gpio_index, prop_values):
     index = 0
     map_array = []
 
@@ -366,6 +366,15 @@ def gpio_map_lookup(cell_parent, gpio_index):
         map_array.append(raw_map[index:index+array_cell_size])
         index += array_cell_size
 
+    # if prop_values, then update the # of parent_specifier_size
+    if prop_values != None:
+        index = 0
+        while index < parent_specifier_size:
+            # assign prop_value skipping controller phandle
+            prop_values[index + 1] = map_array[gpio_index][child_specifier_size + 1 + index]
+            index += 1
+        print("AFTER ", prop_values)
+
     # Return the gpio controller phandle matching the provided index
     return phandles[map_array[gpio_index][child_specifier_size]]
 
@@ -393,7 +403,7 @@ def extract_controller(node_path, prop, prop_values, index,
             # parent is a 'gpio nexus node' (ie has gpio-map)
             try:
                 # controller is to be found in the map, using elem[1] as index
-                cell_parent = gpio_map_lookup(cell_parent, elem[1])
+                cell_parent = gpio_map_lookup(cell_parent, elem[1], prop_values)
             except IndexError:
                 raise Exception("GPIO index not available in gpio-map")
 
@@ -459,7 +469,7 @@ def extract_cells(node_path, prop, prop_values, names, index,
             # parent is a 'gpio nexus node' (ie has gpio-map)
             try:
                 # controller is to be found in the map, using elem[1] as index
-                cell_parent = gpio_map_lookup(cell_parent, elem[1])
+                cell_parent = gpio_map_lookup(cell_parent, elem[1], None)
             except IndexError:
                  raise Exception("GPIO index not available in gpio-map")
 ```

And ended up with:
```
#define DT_NXP_KINETIS_UART_4006D000_UBLOX_SARA_R4_4006D000_SARA_R4_MDM_POWER_GPIOS_CONTROLLER		"GPIO_0"
#define DT_NXP_KINETIS_UART_4006D000_UBLOX_SARA_R4_4006D000_SARA_R4_MDM_POWER_GPIOS_FLAGS		0
#define DT_NXP_KINETIS_UART_4006D000_UBLOX_SARA_R4_4006D000_SARA_R4_MDM_POWER_GPIOS_PIN			2
#define DT_NXP_KINETIS_UART_4006D000_UBLOX_SARA_R4_4006D000_SARA_R4_MDM_RESET_GPIOS_CONTROLLER		"GPIO_2"
#define DT_NXP_KINETIS_UART_4006D000_UBLOX_SARA_R4_4006D000_SARA_R4_MDM_RESET_GPIOS_FLAGS		0
#define DT_NXP_KINETIS_UART_4006D000_UBLOX_SARA_R4_4006D000_SARA_R4_MDM_RESET_GPIOS_PIN			2
```

erwango commented:
@hashcode, Txs for the info, it worked initially, but it seems I've not tested the refactoring enough. Let me rework that.

erwango commented:
@ulfalizer, here is the last batch, which has the merit of fixing bug reported by @Hashcode (so slightly more complex treatment) and I hope fixing your comments (hence somewhat more readable).

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -350,6 +398,15 @@ def extract_controller(node_path, prop, prop_values, index,
+            # Parent is a gpio 'nexus node' (ie has gpio-map).
+            # Controller should be found in the map, using elem[1] as index.
+            # Pin attribues (number, flag) will not be used in this function
+            cell_parent, unused = child_to_parent_unmap(cell_parent, elem[1])
```
Call it `_` instead of `unused`. That's common in Python, and `pylint` knows not to flag it.

erwango replied:
ok, thanks for the hint.

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,54 @@ def build_cell_array(prop_array):
+    # cell_parent provided (which has been identified as a 'nexus node',
+    # ie: has a 'gpio-map' property) thanks to the gpio_index provided.
+    # Note: Nexus nodes and gpio-map property are described in devide tree
+    # specification, v0.3, chapter 'Nexus nodes and Specifier Mapping'.
```
Is "v0.3" a typo? Can't find that version online.

erwango replied:
It's available here: https://github.com/devicetree-org/devicetree-specification/blob/4b1dac80eaca45b4babf5299452a951008a5d864/source/devicetree-basics.rst#nexus-nodes-and-specifier-mapping.
"Soon to be released" according to maintainer.

ulfalizer replied:
Sure it'll be called v0.3?

erwango replied:
Current one is v0.2, so I guess this is a reasonable assumption (https://www.devicetree.org/)

ulfalizer replied:
Bad idea to refer to a v0.3 version that's not out and not sure to be called v0.3 I think. Could say something like "in the upcoming (presumably v0.3) Device Tree specification" instead.

Small typo: devide -> device

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,54 @@ def build_cell_array(prop_array):
     return ret_array
 
+def child_to_parent_unmap(cell_parent, gpio_index):
+    # Aim of this function is to return a gpio-controller from the
```
More of a comment, but instead of "aim of this function...", it might be better to say what the function does.

"This function takes a foo and a bar and returns a (baz, qaz) tuple. If foo can't be found, this other value is returned instead", etc.

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,54 @@ def build_cell_array(prop_array):
+    # Assumption 1: We assume parent #gpio-cells is constant across
+    # the map, so we take the value of the first occurrence and apply
+    # to the whole map.
+    parent = phandles[child_specifier_size+1]
```
I don't get this part. How is `child_specifier_size` ("The number of 32-bit cells required to specify this component", according to the DT spec) related to phandles?

erwango replied:
gpio-map is encoded in the following way:
```
        gpio-map = <0 0 &soc_gpio1 1 0>,                   
                   <1 0 &soc_gpio2 4 0>,
                   <2 0 &soc_gpio1 3 0>,
                   <3 0 &soc_gpio2 2 0>;
```
Taking the first line:
0 0: Child specifier (pin number, pin flag)
&soc_gpio2: GPIO controller phandle
1 0: Parent specifier (pin number, pin flag)

At this point we're looking for GPIO controller phandle which is the element number 'child_specifier_size+1' in the line.

About gpio-map pin flags (does not relates directly to your question, but this might be required for the full comprehension of the change):
Assuming optional params `gpio-map-mask` and `gpio-map-pass-thru` are omitted, child and parent specifier pin flags are not taken into account and pin flag used is the one provided in calling node (typically used in shield overlay file), which is the behavior we expect, at least in current shield use case. This is "Assumption 2".

ulfalizer replied:
AFAICS, this is looking at `#gpio-cells` though. It's saying "the number of GPIO cells + 1 is a phandle". Seems weird to me.

Sorry if I'm missing something.

erwango replied:
So:
0 0: Child specifier (pin number, pin flag) -> length provided by current node's #gpio-cell
&soc_gpio2: GPIO controller phandle
1 0: Parent specifier (pin number, pin flag)  -> length provided by GPIO controller's #gpio-cell

ulfalizer replied:
You're never indexing into `gpio-map` though. You're taking `#gpio-cells + 1` and directly treating that as a phandle.

erwango replied:
Indeed, Txs!

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,56 @@ def build_cell_array(prop_array):
+    parent_specifier_size = reduced[parent]['props']['#gpio-cells']
+    array_cell_size = child_specifier_size + 1 + parent_specifier_size
+
+    # Now that raws length is known, build the map as an array of arrays
```
"Now that the length of each entry in 'gpio-map' is known..." maybe.

erwango replied:
Done

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,56 @@ def build_cell_array(prop_array):
+
+    # Now that raws length is known, build the map as an array of arrays
+    index = 0
+    map_array = []
```
Do you think this could be called `gpio_map_entries` or the like?

erwango replied:
Done

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,56 @@ def build_cell_array(prop_array):
+    # 'Nexus nodes and Specifier Mapping'.
+
+    # First, retrieve gpio-map as a list
+    raw_map = reduced[cell_parent]['props']['gpio-map']
```
Do you think this could be called `gpio_map`?

erwango replied:
It could

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,56 @@ def build_cell_array(prop_array):
+    index = 0
+    map_array = []
+    while index < len(raw_map):
+        map_array.append(raw_map[index:index+array_cell_size])
```
...then this would read as

```
gpio_map_entries.append(gpio_map[index:index+array_cell_size])
```

erwango replied:
yup

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,56 @@ def build_cell_array(prop_array):
+        return None, None
+    else:
+        # Return gpio_controller phandle and specifier pin
+        parent_controller_phandle = map_array[gpio_index][child_specifier_size]
```
Is this looking up the right element?

Note that it's not comparing `gpio_index` against `item[0]` for each `item` in `map_array`, unlike the code above it.

Could maybe do something like this:

```python
for entry in gpio_map_entries:
    if entry[0] == gpio_index:
        parent_controller_phandle = entry[child_specifier_size]
        ...
        return phandles[...]

# gpio_index does not appear in gpio-map
return None, None
```

erwango replied:
Right, I kinda assumed first cell of each raw would be same as raw index.
This is indeed a shortcut, as it could be designed otherwise.

erwango replied:
Fixed

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,58 @@ def build_cell_array(prop_array):
+    index = 0
+    gpio_map_entries = []
+    while index < len(gpio_map):
+        gpio_map_entries.append(gpio_map[index:index+array_cell_size])
```
Could get rid of `gpio_map_entries` at this point, and do this:

```python
for i in range(0, len(gpio_map), array_cell_size):
    entry = gpio_map[i:i+array_cell_size]
    if entry[0] == gpio_index:
        ...
        return ...

# gpio_index did not match any entry in gpio-map
return None, None
```

erwango replied:
Done, and reworked comments

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,52 @@ def build_cell_array(prop_array):
+            # Return gpio_controller and specifier pin
+            return phandles[parent_controller_phandle], parent_pin_number
+
+    # gpio_index did not match any entry in gpio_map_entries
```
Forgot to update this comment.

erwango replied:
Indeed, fixed

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,52 @@ def build_cell_array(prop_array):
+    # First, retrieve gpio-map as a list
+    gpio_map = reduced[cell_parent]['props']['gpio-map']
+
+    # Before parsing, we need to know raws size.
```
Still says "raws". Could say "the size of each entry in 'gpio-map'" or the like.

ulfalizer replied:
Ohh... maybe 'raws' is a typo for 'rows'. I kept reading it as something "raw".

"Before parsing, we need to know the row size in 'gpio-map'", maybe.

erwango replied:
Indeed, fixed

erwango committed:
scripts/dts: globals: parse gpio-map properties for connector nodes

gpio-map is a property of "nexus node", defined in dts v0.3.
It allows to describe a pin connector so it can be referenced
through phandles and hence used in expansion device nodes like a
shield header (typically implemented through overlays).

This change implements gpio controller resolution through these maps.
Few assumptions were taken in order to simplify the implementation.
These assumptions bring some limitations to the use of gpio-map
but my understanding is that this should still allow to cover most
use cases.

Assumptions:
-gpio-size is the same for all gpio-controllers referenced in a map
-optional properties gpio-map-mask and gpio-map-pass-thru are
supposed to be omitted
The understanding of this last assumption is that flags provided in
the expansion device node will overwrite the connector flags.

In a latter stage, when need happen, these limitations can be
revisited to unlock fully fledged gpio-map usage.

Fixes #15637

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,52 @@ def build_cell_array(prop_array):
+
+        if entry[0] == gpio_index:
+            parent_controller_phandle = entry[child_specifier_size]
+            # Assumption 2: We assume otionnal properties 'gpio-map-mask'
```
Fixed, txs

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,52 @@ def build_cell_array(prop_array):
+    gpio_map = reduced[cell_parent]['props']['gpio-map']
+
+    # Before parsing, we need to know 'gpio-map' row size
+    # gpio-map raws are encoded as follows:
```
*rows

ulfalizer reviewed (approved):
Code nits fixed (except one `raws` -> `rows` typo).

Not an expert on how this is actually used, so someone else also ought to check.

Thanks for putting up with the nitting.

ulfalizer commented on scripts/dts/extract/globals.py:
```diff
@@ -331,6 +331,52 @@ def build_cell_array(prop_array):
 
     return ret_array
 
+def child_to_parent_unmap(cell_parent, gpio_index):
```
"Unmap"?

erwango replied:
I felt child_to_parent was not enough specific.
extract?

ulfalizer replied:
`gpio_map_index_to_node` or something maybe (wonder if `index` could be `id` too).

Maybe `cell_parent` could just be called `node_path`, since it's not important to the functionality that it's a parent of something that you're passing in.

erwango replied:
`gpio_map_child_to_parent` ?

erwango commented:
> Thanks for putting up with the nitting.

Thanks for the in deep review

erwango commented:
@mike-scott, updated  https://github.com/erwango/zephyr/tree/test_14057_gpio_map with latest changes

mike-scott commented:
> @mike-scott, updated https://github.com/erwango/zephyr/tree/test_14057_gpio_map with latest changes

Thanks!  I'll try again in a bit.

mike-scott commented:
@erwango nice job, newest change fixes the pin assignment.

mike-scott commented:
@MaureenHelm Can you take a look as I believe MCR20A and possibly other NXP boards have the potential to be used as shields including GPIO pin header assignments.

MaureenHelm commented:
> @MaureenHelm Can you take a look as I believe MCR20A and possibly other NXP boards have the potential to be used as shields including GPIO pin header assignments.

Yes, MCR20A needs this feature to be converted to a shield (#7462). There is also an NXP sensor shield (#10198). I haven't studied this PR in detail yet, but at first glance it looks great.

mike-scott reviewed (approved):
LGTM
